src/Temporalio/Bridge/Worker.cs

```diff
@@ -0,0 +1,335 @@
+        {
+            using (var scope = new Scope())
+            {
+                var completion = new TaskCompletionSource<bool>();
+                unsafe
+                {
+                    Interop.Methods.worker_complete_activity_task(
+                        Ptr,
+                        scope.ByteArray(comp.ToByteArray()),
+                        null,
+                        scope.FunctionPointer<Interop.WorkerCallback>(
+                            (userData, fail) =>
+                            {
+                                if (fail != null)
+                                {
+                                    completion.TrySetException(new InvalidOperationException(
+                                        new ByteArray(Runtime, fail).ToUTF8()));
+                                }
+                                else
+                                {
+                                    completion.TrySetResult(true);
+                                }
+                            }));
+                }
+                await completion.Task.ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Record an activity heartbeat.
+        /// </summary>
+        /// <param name="heartbeat">Heartbeat to record.</param>
+        public void RecordActivityHeartbeat(Api.ActivityHeartbeat heartbeat)
+        {
+            using (var scope = new Scope())
+            {
+                unsafe
+                {
+                    var fail = Interop.Methods.worker_record_activity_heartbeat(
+                        Ptr, scope.ByteArray(heartbeat.ToByteArray()));
+                    if (fail != null)
+                    {
+                        string failStr;
+                        using (var byteArray = new ByteArray(Runtime, fail))
+                        {
+                            failStr = byteArray.ToUTF8();
+                        }
+                        throw new InvalidOperationException(failStr);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Request workflow eviction.
+        /// </summary>
+        /// <param name="runID">Run ID of the workflow to evict.</param>
+        public void RequestWorkflowEviction(string runID)
+        {
+            using (var scope = new Scope())
+            {
+                unsafe
+                {
+                    Interop.Methods.worker_request_workflow_eviction(
+                        Ptr, scope.ByteArray(runID));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Shutdown this worker.
+        /// </summary>
+        /// <returns>Completion task.</returns>
+        public async Task ShutdownAsync()
+        {
+            using (var scope = new Scope())
+            {
+                var completion = new TaskCompletionSource<bool>();
+                unsafe
+                {
+                    Interop.Methods.worker_shutdown(
```

Comment:
For the way I have things ordered, I don't want this to return until pollers are shutdown. Finalize I have as a separate thing just for reclaiming the arc.

Reply:
I'll read the activity worker implementation and reply

Reply:
I read it.
I like that the activity shutdown sequence is contained in the activity worker and doesn't make assumptions about when it is called.
I'd prefer to initiate shutdown externally in TemporalWorker and wait for poller shutdown and completion of all activities in the ActivityWorker.

Reply:
>  I'd prefer to initiate shutdown externally in TemporalWorker and wait for poller shutdown and completion of all activities in the ActivityWorker.

I do initiate shutdown in the TemporalWorker, but yes it is the activity worker caller that waits for poller shutdown and activity completion.

Reply:
When I changed from `tasks.Add(Task.Run(BridgeWorker.ShutdownAsync, CancellationToken.None));` to `BridgeWorker.InitiateShutdown();` (which calls worker `initiate_shutdown`), my `ExecuteActivityAsync_MaxConcurrent_TimesOutIfMore` test fails. This is because once I hit max concurrent and the workflow fails timing out while scheduling the next task, I cancel the running of the worker (i.e. user-defined task completes). But for whatever reason, Core is not sending a `PollActivityError::ShutDown` to my poller. But with `shutdown` it properly does.

Can you test and confirm this scenario? I want the poller to be told it is done by Core on shutdown.